File: CSharp/SpaceVIL/SpaceVIL/Layouts/ItemsLayoutBox.cs
using System;
using System.Collections.Generic;
using System.Linq;
using SpaceVIL.Core;

namespace SpaceVIL
{
    /// <summary>
    /// ItemsLayoutBox is a storage-class that provides an access to existing items.
    /// </summary>
    public static class ItemsLayoutBox
    {
        /// <summary>
        /// Getting existing static items in specified window by its GUID. 
        /// Static items are items that depend on their parent. Parent controls their size, position and etc.
        /// </summary>
        /// <param name="id">GUID of the window.</param>
        /// <returns>The list of existing static items in specified window by its GUID.</returns>
        public static List<IBaseItem> GetLayoutItems(Guid id)
        {
            // return layouts[id].Items.Concat(layouts[id].FloatItems).ToList();
            return layouts[id].Items;
        }
        
        /// <summary>
        /// Getting existing float items in specified window by its GUID. 
        /// Floating items are independent items that do not have a parent, or their root parent is a floating item. 
        /// Examples: SideArea, FloatItem, ContextMenu and etc.
        /// </summary>
        /// <param name="id">GUID of the window.</param>
        /// <returns>The list of existing float items in specified window by its GUID.</returns>
        public static List<IBaseItem> GetLayoutFloatItems(Guid id)
        {
            // return layouts[id].Items.Concat(layouts[id].FloatItems).ToList();
            return layouts[id].FloatItems;
        }


        internal static ItemsLayout GetLayout(Guid id)
        {
            return layouts[id];
        }

        static private Dictionary<Guid, ItemsLayout> layouts = new Dictionary<Guid, ItemsLayout>();

        internal static void InitLayout(Guid _layout)
        {
            ItemsLayout l = new ItemsLayout(_layout);
            layouts.Add(l.Id, l);
        }

        /// <summary>
        /// Adding an item to global item storage (ItemsLayoutBox). 
        /// In usual situation you do not need to use this function only if you create your own implementation of IBaseItem or 
        /// create a new implementation of IFloatItem.
        /// </summary>
        /// <param name="window">Any CoreWindow instance.</param>
        /// <param name="item">Any IBaseItem instance.</param>
        /// <param name="type">Type of an item: SpaceVIL.Core.LayoutType.Static 
        /// or SpaceVIL.Core.LayoutType.Floating.</param>
        public static void AddItem(CoreWindow window, IBaseItem item, LayoutType type)
        {
            switch (type)
            {
                case LayoutType.Static:
                    layouts[window.GetWindowGuid()].Items.Add(item);
                    break;
                case LayoutType.Floating:
                    {
                        layouts[window.GetWindowGuid()].FloatItems.Add(item);
                        item.SetHandler(window);
                        break;
                    }
                case LayoutType.Dialog:
                    layouts[window.GetWindowGuid()].DialogItems.Add(item);
                    break;
                default:
                    layouts[window.GetWindowGuid()].Items.Add(item);
                    break;
            }
        }

        /// <summary>
        /// Removing an item from global item storage (ItemsLayoutBox). 
        /// In usual situation you do not need to use this function only if you create your own implementation of IBaseItem or 
        /// want to remove IFloatItem instance.
        /// </summary>
        /// <param name="window">Any CoreWindow instance.</param>
        /// <param name="item">Any IBaseItem instance.</param>
        /// <param name="type">Type of an item: SpaceVIL.Core.LayoutType.Static 
        /// or SpaceVIL.Core.LayoutType.Floating</param>
        /// <returns>True: if removal was successfull. False: if the specified item does not exist in the storage.</returns>
        public static bool RemoveItem(CoreWindow window, IBaseItem item, LayoutType type)
        {
            window.FreeVRAMResource(item);
            
            switch (type)
            {
                case LayoutType.Static:
                    return layouts[window.GetWindowGuid()].Items.Remove(item);
                // break;
                case LayoutType.Floating:
                    {
                        UnsubscribeWindowSizeMonitoring(item, GeometryEventType.ResizeWidth);
                        UnsubscribeWindowSizeMonitoring(item, GeometryEventType.ResizeHeight);
                        return layouts[window.GetWindowGuid()].FloatItems.Remove(item);
                    }
                // break;
                case LayoutType.Dialog:
                    return layouts[window.GetWindowGuid()].DialogItems.Remove(item);
                // break;
                default:
                    return layouts[window.GetWindowGuid()].Items.Remove(item);
                    // break;
            }
        }

        internal static void SubscribeWindowSizeMonitoring(IBaseItem item, GeometryEventType type)
        {
            //подписка
            item.SetParent(item.GetHandler().GetLayout().GetContainer());
            item.GetHandler().GetLayout().GetContainer().AddEventListener(type, item);
        }
        internal static void UnsubscribeWindowSizeMonitoring(IBaseItem item, GeometryEventType type)
        {
            //отписка
            item.SetParent(null);
            item.GetHandler().GetLayout().GetContainer().RemoveEventListener(type, item);
        }

        /// <summary>
        /// Getting the list of names of existing items in the specified window.
        /// </summary>
        /// <param name="window">Any CoreWindow instance.</param>
        /// <returns>The list of names of existing items in the specified window.</returns>
        public static List<string> GetListOfItemsNames(CoreWindow window) => 
            layouts[window.GetWindowGuid()].Items.Select(_ => _.GetItemName()).ToList();

        internal static string[] GetListOfItemsColors(CoreWindow layout) => 
            layouts[layout.GetWindowGuid()].Items.Select(_ => _.GetBackground().ToString()).ToArray();
        
        /// <summary>
        /// Printing all existing items in the specified window.
        /// </summary>
        /// <param name="window">Any CoreWindow instance.</param>
        public static void PrintListOfItems(CoreWindow window)
        {
            List<string> list = GetListOfItemsNames(window);
            foreach (var item in list)
            {
                Console.WriteLine(item);
            }
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/ItemsLayoutBox.java
package com.spvessel.spacevil;

import com.spvessel.spacevil.Core.IBaseItem;
import com.spvessel.spacevil.Flags.GeometryEventType;
import com.spvessel.spacevil.Flags.LayoutType;

import java.util.*;

/**
 * ItemsLayoutBox is a storage-class that provides an access to existing items.
 */
public class ItemsLayoutBox {
    private ItemsLayoutBox() {

    }

    /**
     * Getting existing static items in specified window by its GUID. 
     * Static items are items that depend on their parent. Parent controls their size, position and etc.
     * @param id UUID of the window.
     * @return The list of existing static items in specified window by its UUID.
     */
    static public List<IBaseItem> getLayoutItems(UUID id) {
        // return layouts[id].Items.Concat(layouts[id].FloatItems).ToList();
        return layouts.get(id).getItems();
    }

    /**
     * Getting existing float items in specified window by its UUID. 
     * Floating items are independent items that do not have a parent, or their root parent is a floating item. 
     * Examples: SideArea, FloatItem, ContextMenu and etc.
     * @param id UUID of the window.
     * @return The list of existing float items in specified window by its UUID.
     */
    static public List<IBaseItem> getLayoutFloatItems(UUID id) {
        // return layouts[id].Items.Concat(layouts[id].FloatItems).ToList();
        return layouts.get(id).getFloatItems();
    }

    static List<IBaseItem> getLayoutDialogItems(UUID id) {
        // return layouts[id].Items.Concat(layouts[id].FloatItems).ToList();
        return layouts.get(id).getDialogItems();
    }

    static ItemsLayout getLayout(UUID id) {
        return layouts.get(id);
    }

    static private Map<UUID, ItemsLayout> layouts = new HashMap<>();

    static void initLayout(UUID _layout) {
        ItemsLayout l = new ItemsLayout(_layout);
        layouts.put(l.getId(), l);
    }

    /**
     * Adding an item to global item storage (ItemsLayoutBox). 
     * In usual situation you do not need to use this function only if you create your own 
     * implementation of IBaseItem or create a new implementation of IFloatItem.
     * @param layout Any CoreWindow instance.
     * @param item Any IBaseItem instance.
     * @param type Type of an item: com.spvessel.spacevil.Flags.LayoutType.STATIC 
     * or com.spvessel.spacevil.Flags.LayoutType.FLOATING.
     */
    static public void addItem(CoreWindow layout, IBaseItem item, LayoutType type) {
        switch (type) {
        case Static:
            layouts.get(layout.getWindowGuid()).getItems().add(item);
            break;
        case Floating: {
            layouts.get(layout.getWindowGuid()).getFloatItems().add(item);
            item.setHandler(layout);
            break;
        }
        case Dialog:
            layouts.get(layout.getWindowGuid()).getDialogItems().add(item);
            break;
        default:
            layouts.get(layout.getWindowGuid()).getItems().add(item);
            break;
        }
    }

    /**
     * Removing an item from global item storage (ItemsLayoutBox). 
     * In usual situation you do not need to use this function only if you create your own 
     * implementation of IBaseItem or want to remove IFloatItem instance.
     * @param layout Any CoreWindow instance.
     * @param item Any IBaseItem instance.
     * @param type Type of an item: com.spvessel.spacevil.Flags.LayoutType.STATIC 
     * or com.spvessel.spacevil.Flags.LayoutType.FLOATING.
     * @return True: if removal was successfull. False: if the specified item does not exist in the storage.
     */
    static public boolean removeItem(CoreWindow layout, IBaseItem item, LayoutType type) {
        layout.freeVRAMResource(item);

        switch (type) {
        case Static:
            return layouts.get(layout.getWindowGuid()).getItems().remove(item);
        // break;
        case Floating: {
            unsubscribeWindowSizeMonitoring(item, GeometryEventType.ResizeWidth);
            unsubscribeWindowSizeMonitoring(item, GeometryEventType.ResizeHeight);
            return layouts.get(layout.getWindowGuid()).getFloatItems().remove(item);
        }
        // break;
        case Dialog:
            return layouts.get(layout.getWindowGuid()).getDialogItems().remove(item);
        // break;
        default:
            return layouts.get(layout.getWindowGuid()).getItems().remove(item);
        // break;
        }
    }

    static void subscribeWindowSizeMonitoring(IBaseItem item, GeometryEventType type) {
        // подписка
        item.setParent(item.getHandler().getLayout().getContainer());
        item.getHandler().getLayout().getContainer().addEventListener(type, item);
    }

    static void unsubscribeWindowSizeMonitoring(IBaseItem item, GeometryEventType type) {
        // отписка
        item.setParent(null);
        item.getHandler().getLayout().getContainer().removeEventListener(type, item);
    }

    /**
     * Getting the list of names of existing items in the specified window.
     * @param layout Any CoreWindow instance.
     * @return The list of names of existing items in the specified window.
     */
    static public String[] getListOfItemsNames(CoreWindow layout) {
        String[] result = new String[layouts.get(layout.getWindowGuid()).getItems().size()];
        for (int i = 0; i < result.length; i++) {
            result[i] = layouts.get(layout.getWindowGuid()).getItems().get(i).getItemName();
        }
        return result;
    }

    static String[] getListOfItemsColors(CoreWindow layout) {
        String[] result = new String[layouts.get(layout.getWindowGuid()).getItems().size()];
        for (int i = 0; i < result.length; i++) {
            result[i] = layouts.get(layout.getWindowGuid()).getItems().get(i).getBackground().toString();
        }
        return result;
    }

    /**
     * Printing all existing items in the specified window.
     * @param layout Any CoreWindow instance.
     */
    static public void printListOfItems(CoreWindow layout) {
        String[] list = getListOfItemsNames(layout);
        for (String item : list) {
            System.out.println(item);
        }
    }
}